### ISSUE TITLE
[Alt Gr] doesn't work with -K parameter

### ISSUE THREAD

**Issue by Manu99it** [Genymobile/scrcpy#4948]
- [x] I have read the [FAQ](https://github.com/Genymobile/scrcpy/blob/master/FAQ.md).
 - [x] I have searched in existing [issues](https://github.com/Genymobile/scrcpy/issues).

**Environment**
 - OS: Windows
 - scrcpy version: 2.4
 - installation method: extracted zip win release
 - device model: Pixel 5 (google)
 - Android version: 14

**Describe the bug**
I installed this app https://github.com/calin-darie/extra-keyboard-layouts to have the UK extended layout I use.
On Windows if you hold [Alt Gr] and type a letter you can use the acute accent: é, á, í, ó, ú, ý. It doesn't behave like a dead key because you need to hold it. However on android [Alt Gr] + [vowel] does nothing. It's an Android limitation? Or app limitation?  I am posting the issue here because I use the keyboard through scrcpy, and I can't test it directly on android because it has a proprietary connector (surface keyboard). If by scrcpy all is done correctly I'll post an issue on the app git. When using UK extended with italian it's needed for example, especially the 'é'. Also [Alt Gr] + backtick (grave accent) on windows produces ¦ and similarly [Alt Gr] + 4 produces the eur symbol €. On Android nothing happens, at least through scrcpy

**Comment by rom1v** [comment]
> On Windows if you hold [Alt Gr] and type a letter you can use the acute accent: é, á, í, ó, ú, ý

It is probably a Windows-specific behavior. Not all OS behave the same way (it is not a "limitation", it's just different dehaviors).

> and similarly [Alt Gr] + 4 produces the eur symbol €

<kbd>AltGr</kbd>+<kbd>e</kbd> produces € (at least with a French layout).

**Comment by Manu99it** [comment]
@rom1v so I managed to try with a true physical keyboard and AltGr does in fact works correctly and behaves the same as Windows with the UK extended layout or any other layout. I think that when there is uhid setting enabled, scrcpy doesn't pass the AltGr key. Alt is the key you use for shortcuts, maybe it goes in conflict

**Comment by rom1v** [comment]
What keyboard layout do you use on Android for UHID?

(press <kbd>MOD</kbd>+<kbd>k</kbd> while mirroring with UHID is enabled: https://github.com/Genymobile/scrcpy/blob/master/doc/keyboard.md#physical-keyboard-simulation)

**Comment by Manu99it** [comment]
> What keyboard layout do you use on Android for UHID?
> 
> (press MOD+k while mirroring with UHID is enabled: https://github.com/Genymobile/scrcpy/blob/master/doc/keyboard.md#physical-keyboard-simulation)

As I said, I installed the app I mentioned that adds UK extended layout. But AltGr for me didn't work with ANY layout (I tried Italian (my language), France, etc. AltGr doesn't work)

**Comment by rom1v** [comment]
Indeed, if you press a modifier key, it only set the modifier in the HID protocol (so it is taken into account for all my key combinations), but the scancode of the modifier key itself is never sent. Run with `scrcpy -K -Vverbose --no-audio`:

```
VERBOSE: hid keyboard: key down scancode=e6 (230) mod=40
VERBOSE: input: UHID input [1] 40 00 00 00 00 00 00 00 
VERBOSE: hid keyboard: key up   scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
```

That may be important for your use case.

Please test with this quick&dirty workaround:

```diff
diff --git a/app/src/hid/hid_keyboard.h b/app/src/hid/hid_keyboard.h
index ddd2cc91c..b5ffff04d 100644
--- a/app/src/hid/hid_keyboard.h
+++ b/app/src/hid/hid_keyboard.h
@@ -12,7 +12,7 @@
 // Maybe SDL_Keycode is used by most people, but SDL_Scancode is taken from USB
 // HID protocol.
 // 0x65 is Application, typically AT-101 Keyboard ends here.
-#define SC_HID_KEYBOARD_KEYS 0x66
+#define SC_HID_KEYBOARD_KEYS 0xFF
 
 extern const uint8_t SC_HID_KEYBOARD_REPORT_DESC[];
 extern const size_t SC_HID_KEYBOARD_REPORT_DESC_LEN;
```

With it, the scancode is sent:

```
VERBOSE: hid keyboard: key down scancode=e6 (230) mod=40
VERBOSE: input: UHID input [1] 40 00 E6 00 00 00 00 00 
VERBOSE: keys[e6] = false
VERBOSE: hid keyboard: key up   scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
VERBOSE: keys[e2] = false
```

**Comment by Manu99it** [comment]
Now that I see it in the debugger, even with the old server scrcpy IS passing the AltGr key, right? Maybe it passes Alt_R instead of AltGr though

**Comment by rom1v** [comment]
Sorry, I posted a server binary whereas the change was in the client 🤦 (doing several thing at once…)

Now I'm on the phone so I will post a binary later.

**Comment by rom1v** [comment]
Here is a binary:

- [`scrcpy-win64-issue4948.zip`](https://tmp.rom1v.com/scrcpy/4948/1/scrcpy-win64-issue4948.zip) <sub>`SHA-256: d9bdb90548b1ba3a93786fca32151ccbe8afe84226ab270375b0cb1fccbe312`</sub>

**Comment by Manu99it** [comment]
This is the output. The AltGr key still seems to be not functioning
![image](https://github.com/Genymobile/scrcpy/assets/19348772/12860d96-1c99-439e-a236-fcd650c5e94e)

**Comment by rom1v** [comment]
OK, so this scancode is not the source of the difference with the behavior you get with a keyboard.

Could you test with AOA instead of UHID? https://github.com/Genymobile/scrcpy/blob/master/doc/keyboard.md#aoa

**Comment by Manu99it** [comment]
It doesn't work either. Here is the console debug 
![image](https://github.com/Genymobile/scrcpy/assets/19348772/b26154de-e5a7-4681-a85d-cd7a6d18328e)

**Comment by Manu99it** [comment]
I don't understand why, but seems also dead keys aren't working with UHID, although yesterday it seemed so. With OTG/aoa dead keys do work

**Comment by Manu99it** [comment]
I found a solution: holding AltGr, then pressing and releasing LCtrl does input the acute accent! I noted that scrcpy sends 2 keyboard events instead of one when pressing altgr, pressinf LCtrl "deletes" that one and then AltGr works correctly

**Comment by Manu99it** [comment]
@rom1v I've scrolled back a few issues and noted #4725 so probably my issue is a duplicate of that one. And in AOA mode the behaviour is the same. If the sdl library won't patch this you could add a workaround for windows or hint windows users to pressing and releasing LCtrl while holding altgr for the correct behaviour on android

**Comment by neclepsio** [comment]
> I found a solution: holding AltGr, then pressing and releasing LCtrl does input the acute accent! I noted that scrcpy sends 2 keyboard events instead of one when pressing altgr, pressing LCtrl "deletes" that one and then AltGr works correctly

Thank you! Maybe this workaround should be documented until SDL update?

**Comment by xlotlu** [comment]
On Linux the RAlt event is sent on its own, but for some reason it's not processed as AltGr on the other side. Here's a standalone key press:

```
VERBOSE: hid keyboard: key down scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
VERBOSE: hid keyboard: key up   scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
```

Now, trying to enter a "`ș`", but only getting an "`s`":

```
VERBOSE: hid keyboard: key down scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
VERBOSE: keys[16] = true
VERBOSE: hid keyboard: key down scancode=16 (22) mod=00
VERBOSE: input: UHID input [1] 00 00 16 00 00 00 00 00 
VERBOSE: keys[16] = false
VERBOSE: hid keyboard: key up   scancode=16 (22) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
VERBOSE: hid keyboard: key up   scancode=e6 (230) mod=00
VERBOSE: input: UHID input [1] 00 00 00 00 00 00 00 00 
```

I'm using the same "More Physical Keyboard Layouts" app, with a "Programmers" layout (which is a regular US layout, with diacritics on their respective keys when using AltGr). The result is I cannot input diacritics.

Without the `-K` switch, one gets 

```
VERBOSE: input: key down code=47 repeat=0 meta=000000
VERBOSE: input: text "ș"
[server] WARN: Could not inject char u+0219
VERBOSE: input: key up   code=47 repeat=0 meta=000000
```

**Comment by Gattsuu** [comment]
This is still (or again) a problem btw. Luckily the workaround suggested by @Manu99it indeed still works. It is still annoying though. I'm on 2025-05 Cumulative Update for Windows 11 Version 24H2 (KB5058411) trying to control a Samsung s9 tablet if that matters.
